credentials/gcp.go

```diff
@@ -0,0 +1,171 @@
+	ProjectID               string   `json:"project_id"`
+	RedirectUris            []string `json:"redirect_uris"`
+	TokenURI                string   `json:"token_uri"`
+}
+
+// DeleteCredential removes a stored credential from the system keystore, and
+// will remove the assigned defualt credential from configuration.
+func (gc GoogleCredential) DeleteCredential(credentialName string) error {
+	return keyring.Delete("scuttle", credentialName+"-gcp")
+}
+
+// SaveCredential marshals the received GoogleToken and stores the resulting json blob in the
+// system keystore for secure storage at rest.
+func (gc GoogleCredential) SaveCredential(credentialName string, credential GoogleToken) error {
+	jsonData, err := json.Marshal(credential)
+	if err != nil {
+		return err
+	}
+	return keyring.Set("scuttle", credentialName+"-gcp", string(jsonData))
+}
+
+// GetCredential returns a decoded GoogleToken from the system keystore. The resulting
+// object is serializeable and should be used in conjunction with google.configFromJSON()
+func (gc GoogleCredential) GetCredential(credentialName string) (GoogleToken, error) {
+	userTokenJSON, err := keyring.Get("scuttle", credentialName+"-gcp")
+	if err != nil {
+		return GoogleToken{}, err
+	}
+	var token GoogleToken
+	err = json.Unmarshal([]byte(userTokenJSON), &token)
+	if err != nil {
+		return GoogleToken{}, err
+	}
+
+	return token, nil
+}
+
+func (gc GoogleCredential) JSON() ([]byte, error) {
+	// This serializer method handles the override from ENV transparently
+	// to the user. We should signal that we're loading from the ENV var
+	// declared credential.
+	external, exists := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
+	if exists {
+		log.Printf("Using env exported GOOGLE_APPLICATION_CREDENTIALS")
+		f, err := ioutil.ReadFile(external)
+		if err != nil {
+			log.Fatalf("Unable to read GOOGLE_APPLICATION_CREDENTIALS: %v", err)
+		}
+		return f, nil
+	}
+
+	tok, err := gc.GetCredential("default")
+	if err != nil {
+		log.Println("Unable to locate credentials. Have you run `sctl credential add`?")
+		log.Println("Another common issue is if running in a headless environment, where sctl expects" +
+			" to be invoked with GOOGLE_APPLICATION_CREDENTIALS environment variable set.")
+		log.Fatalf("Failed definetly with error %v", err)
+	}
+	return json.Marshal(tok)
+}
+
+// ValidateContext scans the environment for a declaration of authentication via ENV
+// In the case of the GCP provider, we scan for GOOGLE_APPLICAITON_CREDENTIALS and return
+// an error if the environment variable is declared.
+func (gc GoogleCredential) ValidateContext() error {
+	_, isSet := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
+	if isSet {
+		return errors.New("Environment Variable GOOGLE_APPLICATION_CREDENTIALS is present, and will prevent sctl from using any user configured credentials")
+	}
+
+	return nil
+}
+
+func (gc GoogleCredential) formatCredential(token *oauth2.Token, rawClientData []byte) GoogleToken {
```

Comment:
not exported so doesn't require a docstring, but I think its still good practice to include one